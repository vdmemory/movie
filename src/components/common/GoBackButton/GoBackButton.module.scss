.goBack {
    padding: 15px;
    cursor: pointer;
    background: aliceblue;
    text-align: center;
    display: flex;
    justify-content: center;
    align-items: center;
    height: 80px;

    & .iconButton {
        margin-right: 15px;
    }
}