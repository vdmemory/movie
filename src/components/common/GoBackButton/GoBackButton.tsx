import React, { useCallback } from "react";
import { useHistory } from "react-router-dom";
import { HiOutlineArrowLeft } from "react-icons/hi";

import styles from "./GoBackButton.module.scss";

type Props = {
    name: string;
}

const GoBackButton: React.FC<Props> = ({name}) => {
    const history = useHistory();

    const handleGoBack = useCallback(() => {
        history.goBack();
    }, [history]);

    return (
        <div className={styles.goBack} onClick={handleGoBack}>
            <HiOutlineArrowLeft className={styles.iconButton} />
            {name}
        </div>
    );
};

export default GoBackButton;
