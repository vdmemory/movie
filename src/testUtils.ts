export const deepEqual = (a: any, b: any) => {
    if (a === b) {
        return true;
    }

    if (
        a == null ||
        typeof a != "object" ||
        b == null ||
        typeof b != "object"
    ) {
        return false;
    }

    var propertiesInA = 0,
        propertiesInB = 0;
    for (var property in a) {
        propertiesInA += 1;
    }
    for (var property in b) {
        propertiesInB += 1;
        if (!(property in a) || !deepEqual(a[property], b[property])) {
            return false;
        }
    }
    return propertiesInA == propertiesInB;
};
