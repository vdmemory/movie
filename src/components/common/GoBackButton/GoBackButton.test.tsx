import React from "react";
import { createRenderer } from "react-test-renderer/shallow";
import { useHistory } from "react-router-dom";

import { deepEqual } from "../../../testUtils";
import GoBackButton from "./GoBackButton";
import { HiOutlineArrowLeft } from "react-icons/hi";

type Props = {
    name: string;
};

const setup = (props: Props) => {
    const renderer = createRenderer();
    renderer.render(<GoBackButton {...props} />);
    const output = renderer.getRenderOutput();
    return { output, props, renderer };
};

const propsRender: Props = {
    name: "Back To List",
};

jest.mock("react-router-dom", () => {
    const goBackSpy = jest.fn();
    return {
        useHistory: function history() {
            return {
                goBack: goBackSpy,
            };
        },
    };
});

describe("GoBackButton Component", () => {
    describe("Render component and its components", () => {
        test("Should renders correctly", () => {
            const { output } = setup(propsRender);
            expect(output).toMatchSnapshot();
        });

        test("Should render container and has a movie class and check text", () => {
            const { output } = setup(propsRender);
            const [, text] = output.props.children;
            expect(output.type).toBe("div");
            expect(output.props.className).toBe("goBack");
            expect(text).toBe("Back To List");
        });

        test("Should call handleGoBack on click", () => {
            const history = useHistory();
            const { output } = setup(propsRender);
            output.props.onClick();
            expect(history.goBack).toBeCalled();
            // @ts-expect-error history.goBack is mocked
            history.goBack.mockClear();
        });

        describe("HiOutlineArrowLeft", () => {
            test("Should render and check class", () => {
                const { output } = setup(propsRender);
                const [component] = output.props.children;
                expect(component.type).toBe(HiOutlineArrowLeft);
                expect(component.props.className).toBe("iconButton");
            });
        });

        describe("Check props", () => {
            const expectedProps = {
                name: "Back To List",
            };
            test("Should render and check props", () => {
                const { props } = setup(propsRender);
                expect(deepEqual(props, expectedProps)).toBeTruthy();
            });
        });
    });
});
