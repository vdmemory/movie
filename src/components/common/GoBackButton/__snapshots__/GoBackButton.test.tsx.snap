// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GoBackButton Component Render component and its components Should renders correctly 1`] = `
<div
  className="goBack"
  onClick={[Function]}
>
  <HiOutlineArrowLeft
    className="iconButton"
  />
  Back To List
</div>
`;
